contracts/random/RandomSeedProvider.sol

```diff
@@ -0,0 +1,292 @@
+     * @dev Only RANDOM_ADMIN_ROLE can do this.
+     * @param _consumer Game contract that inherits from RandomValues.sol that is no longer authorised to use off-chain random.
+     */
+    function removeOffchainRandomConsumer(address _consumer) external onlyRole(RANDOM_ADMIN_ROLE) {
+        approvedForOffchainRandom[_consumer] = false;
+        emit OffchainRandomConsumerRemoved(_consumer);
+    }
+
+    /**
+     * @notice Request the index number to track when a random number will be produced.
+     * @dev Note that the same _randomFulfillmentIndex will be returned to multiple games and even within
+     * @dev the one game. Games must personalise this value to their own game, the the particular game player,
+     * @dev and to the game player's request.
+     * @return _randomFulfillmentIndex The index for the game contract to present to fetch the next random value.
+     * @return _randomSource Indicates that an on-chain source was used, or is the address of an offchain source.
+     */
+    function requestRandomSeed() external returns (uint256 _randomFulfillmentIndex, address _randomSource) {
+        if (randomSource == ONCHAIN || !approvedForOffchainRandom[msg.sender]) {
+            // Generate a value for this block if one has not been generated yet. This
+            // is required because there may have been calls to requestRandomSeed
+            // in previous blocks that are waiting for a random number to be produced.
+            _generateNextRandomOnChain();
+
+            // Indicate that a value based on the next block will be fine.
+            _randomFulfillmentIndex = nextRandomIndex;
+
+            _randomSource = ONCHAIN;
+        } else {
+            // Limit how often offchain random numbers are requested to a maximum of once per block.
+            if (lastBlockOffchainRequest == block.number) {
+                _randomFulfillmentIndex = prevOffchainRandomRequest;
+            } else {
+                _randomFulfillmentIndex = IOffchainRandomSource(randomSource).requestOffchainRandom();
+                prevOffchainRandomRequest = _randomFulfillmentIndex;
+                lastBlockOffchainRequest = block.number;
+            }
+            _randomSource = randomSource;
+        }
+    }
+
+    /**
+     * @notice Fetches a random seed value that was requested using requestRandom.
+     * @dev Note that the same _randomSeed will be returned to multiple games and even within
+     * @dev the one game. Games must personalise this value to their own game, the the particular game player,
+     * @dev and to the game player's request.
+     * @return _randomSeed The value from with random values can be derived.
+     */
+    function getRandomSeed(
+        uint256 _randomFulfillmentIndex,
+        address _randomSource
+    ) external returns (bytes32 _randomSeed) {
+        if (_randomSource == ONCHAIN) {
+            _generateNextRandomOnChain();
+            if (_randomFulfillmentIndex >= nextRandomIndex) {
```

Comment:
is there a valid scenario where `_randomFulfillmentIndex > nextRandomIndex` or would that always imply an invalid `_randomFullfillmentIndex` value?

Reply:
nextRandomIndex and _randomFulfillmentIndex are related to the block number. If random seeds where requested each block, then they would directly correlate with the block number. Hence, a _randomFulfillmentIndex which is larger than nextRandomIndex indicates that fulfilment will occur in a future block. Returning "wait" via a revert it the best way to handle this situation.

Reply:
Interesting. Given [this assignment  after nextIndex is incremented](https://github.com/immutable/contracts/blob/9813597ceceb196729a22c8b8a7bd25a5179d37a/contracts/random/RandomSeedProvider.sol#L184), it is not immediately clear to me when `_randomFullfilmentIndex` might have a higher value... but I am sure I just missed something

Reply:
Actually, you are not missing something. Historically, the fulfilment index could have been some blocks in the future. The realisation I had was that it could just be one block in the future. Hence, just checking equivalence would be OK.